### PR TITLE
feat(dev-server): add "ping" route

### PR DESCRIPTION
**Pull request by tanner-reits**
## What is the current behavior?

The Stencil dev server will "spin up" prior to the accompanying project build completing. This can cause issues for processes (such as E2E testing tools like Playwright), that rely on the dev server to work correctly. 

GitHub Issue Number: N/A

## What is the new behavior?

A new route is added to the dev server request handler. This route will return a 200 response once the Stencil build is completed and successful. If a build fails, or another error is encountered, the route will return a 500 error.

This route is configurable via the `devServer` object on the Stencil config. If the route is explicitly set to `null`, no route will be registered with the handler. 

## Documentation

https://github.com/ionic-team/stencil-site/pull/1363

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

**Automated tests**

Unit tests were added to the request handler and validation test suites.

**Manual tests**

Functionality can be tested by installing a build of this branch into any Stencil project. After running `npm start`, navigate to `/ping`. Can also test changing the value in the project's config or setting it `null`.

## Other information

This change was made as a part of work to implement a Playwright testing adapter for Stencil. Prior to this change, tests would start execution prematurely, leading to quick failures.

### CODE REVIEW

File: src/declarations/stencil-public-compiler.ts [hunk 1]
```diff
@@ -651,6 +651,15 @@ export interface DevServerConfig extends StencilDevServerConfig {
...
   srcIndexHtml?: string;
+
+  /**
+   * Route to be used for the "ping" sub-route of the Stencil dev server.
+   * This route will return a 200 status code once the Stencil build has finished.
+   * Setting this to `null` will disable the ping route.
+   *
+   * Defaults to `/ping`
+   */
+  pingRoute?: string | null;
```
Comment (tanner-reits) [inline src/declarations/stencil-public-compiler.ts:662]: Went with `pingRoute`, but totally fine if we wanted to go with something like "health" or "status".

File: src/compiler/config/test/validate-dev-server.spec.ts [hunk 2]
```diff
@@ -276,4 +276,23 @@ describe('validateDevServer', () => {
     const { config } = validateConfig(inputConfig, mockLoadConfigInit());
     expect(config.devServer.prerenderConfig).toBe(wwwOutputTarget.prerenderConfig);
   });
+
+  describe('pingRoute', () => {
+    it('should default to /ping', () => {
+      const { config } = validateConfig(inputConfig, mockLoadConfigInit());
```
Comment (rwaskiewicz) [inline src/compiler/config/test/validate-dev-server.spec.ts:282]: Non-blocking - can we future proof this a little and make sure `pingRoute` isn't available on our config? 
```suggestion
      delete inputConfig.devServer.pingRoute;
      const { config } = validateConfig(inputConfig, mockLoadConfigInit());
```
I wouldn't be surprised if we added a default of '/ping' to the `inputDevServerConfig` in the future as a part of another effort and altered how this test is set up accidentally

File: src/compiler/config/test/validate-dev-server.spec.ts [hunk 3]
```diff
@@ -276,4 +276,23 @@ describe('validateDevServer', () => {
...
+      expect(config.devServer.pingRoute).toBe('/ping');
+    });
+
+    it('should set user defined pingRoute', () => {
+      inputConfig.devServer = { ...inputDevServerConfig, pingRoute: '/my-ping' };
+      const { config } = validateConfig(inputConfig, mockLoadConfigInit());
+      expect(config.devServer.pingRoute).toBe('/my-ping');
+    });
+
```
Comment (rwaskiewicz) [inline src/compiler/config/test/validate-dev-server.spec.ts:291]: Can we add a test for `pingRoute`s without a leading '/'?
```suggestion

    it('should prefix pingRoute with a "/"', () => {
      inputConfig.devServer = { ...inputDevServerConfig, pingRoute: 'my-ping' };
      const { config } = validateConfig(inputConfig, mockLoadConfigInit());
      expect(config.devServer.pingRoute).toBe('/my-ping');
    });
```

File: src/declarations/stencil-public-compiler.ts (hunk 1 above)
Comment (rwaskiewicz) [reply src/declarations/stencil-public-compiler.ts:662]: I'm good with `pingRoute` 👍

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]
Two minor asks, only one blocking

**Review by christian-bromann — APPROVED** [review]

**Review by rwaskiewicz — APPROVED** [review]

**Comment by christian-bromann** [comment]
This has been released with Stencil 🚞 [v4.13.0](https://github.com/ionic-team/stencil/releases/tag/v4.13.0)
